## BLOK 5 — Aplikacja użytkowa i operacyjne użycie modelu dziennego

## 5.1 Architektura aplikacji i kontrakt wejścia

In [1]:
# 5.1.1 Definicja architektury aplikacji i kontraktu wejścia

from pathlib import Path
import pandas as pd

# Resolve current working directory for BLOK 5
step5u_11_block5_dir = Path.cwd().resolve()

# Resolve project root by walking up the directory tree
step5u_11_project_root = None

for candidate in [step5u_11_block5_dir] + list(step5u_11_block5_dir.parents):
    if (candidate / "4_Wdrozenie_i_uzycie_modelu").exists():
        step5u_11_project_root = candidate
        break

assert step5u_11_project_root is not None, (
    "Project root could not be resolved from the current working directory."
)

# Define Block 4 dependency paths
step5u_11_block4_dir = step5u_11_project_root / "4_Wdrozenie_i_uzycie_modelu"
step5u_11_input_model_package_dir = step5u_11_block4_dir / "input_model_package"
step5u_11_outputs_dir = step5u_11_block4_dir / "outputs_dzien_stacja"
step5u_11_db_dir = step5u_11_block4_dir / "db_scoring"

# Define Block 5 application paths
step5u_11_app_path = step5u_11_block5_dir / "aplikacja_dzien_stacja.py"
step5u_11_app_contract_artifact_path = step5u_11_outputs_dir / "b5_01_app_contract.json"

# Define required application source paths
step5u_11_model_path = step5u_11_input_model_package_dir / "b4_15_best_model.joblib"
step5u_11_inference_config_path = step5u_11_input_model_package_dir / "b4_15_inference_config.json"
step5u_11_feature_importance_path = step5u_11_input_model_package_dir / "b4_15_feature_importance.parquet"
step5u_11_predictions_for_app_path = step5u_11_outputs_dir / "b4u_05_predictions_for_app.parquet"
step5u_11_app_handoff_path = step5u_11_outputs_dir / "b4u_05_app_handoff.json"

# Build directory inventory
step5u_11_directory_inventory_df = pd.DataFrame(
    [
        {
            "directory_name": "project_root",
            "directory_role": "application_project_root",
            "directory_path": str(step5u_11_project_root),
            "exists": int(step5u_11_project_root.exists()),
            "is_required": 1,
        },
        {
            "directory_name": "block5_dir",
            "directory_role": "application_notebook_location",
            "directory_path": str(step5u_11_block5_dir),
            "exists": int(step5u_11_block5_dir.exists()),
            "is_required": 1,
        },
        {
            "directory_name": "input_model_package",
            "directory_role": "frozen_model_package",
            "directory_path": str(step5u_11_input_model_package_dir),
            "exists": int(step5u_11_input_model_package_dir.exists()),
            "is_required": 1,
        },
        {
            "directory_name": "outputs_dzien_stacja",
            "directory_role": "application_output_layer",
            "directory_path": str(step5u_11_outputs_dir),
            "exists": int(step5u_11_outputs_dir.exists()),
            "is_required": 1,
        },
        {
            "directory_name": "db_scoring",
            "directory_role": "application_runtime_storage",
            "directory_path": str(step5u_11_db_dir),
            "exists": int(step5u_11_db_dir.exists()),
            "is_required": 1,
        },
    ]
)

# Build source inventory
step5u_11_source_inventory_df = pd.DataFrame(
    [
        {
            "source_name": "aplikacja_dzien_stacja.py",
            "source_role": "application_script_target",
            "source_path": str(step5u_11_app_path),
            "exists": int(step5u_11_app_path.exists()),
            "is_required": 0,
        },
        {
            "source_name": "b4_15_best_model.joblib",
            "source_role": "trained_model",
            "source_path": str(step5u_11_model_path),
            "exists": int(step5u_11_model_path.exists()),
            "is_required": 1,
        },
        {
            "source_name": "b4_15_inference_config.json",
            "source_role": "inference_config",
            "source_path": str(step5u_11_inference_config_path),
            "exists": int(step5u_11_inference_config_path.exists()),
            "is_required": 1,
        },
        {
            "source_name": "b4_15_feature_importance.parquet",
            "source_role": "feature_importance_ranking",
            "source_path": str(step5u_11_feature_importance_path),
            "exists": int(step5u_11_feature_importance_path.exists()),
            "is_required": 1,
        },
        {
            "source_name": "b4u_05_predictions_for_app.parquet",
            "source_role": "app_ready_predictions",
            "source_path": str(step5u_11_predictions_for_app_path),
            "exists": int(step5u_11_predictions_for_app_path.exists()),
            "is_required": 1,
        },
        {
            "source_name": "b4u_05_app_handoff.json",
            "source_role": "application_handoff_contract",
            "source_path": str(step5u_11_app_handoff_path),
            "exists": int(step5u_11_app_handoff_path.exists()),
            "is_required": 1,
        },
    ]
)

# Validate required directories and required sources
step5u_11_missing_required_directories = step5u_11_directory_inventory_df.loc[
    (step5u_11_directory_inventory_df["is_required"] == 1)
    & (step5u_11_directory_inventory_df["exists"] == 0),
    "directory_name",
].tolist()
assert not step5u_11_missing_required_directories, (
    f"Missing required directories: {step5u_11_missing_required_directories}"
)

step5u_11_missing_required_sources = step5u_11_source_inventory_df.loc[
    (step5u_11_source_inventory_df["is_required"] == 1)
    & (step5u_11_source_inventory_df["exists"] == 0),
    "source_name",
].tolist()
assert not step5u_11_missing_required_sources, (
    f"Missing required application sources: {step5u_11_missing_required_sources}"
)

# Build architecture summary
step5u_11_architecture_summary_df = pd.DataFrame(
    [
        {
            "app_file_exists": int(step5u_11_app_path.exists()),
            "app_build_target_defined": int(step5u_11_app_path.parent.exists()),
            "required_directory_count": int(
                (step5u_11_directory_inventory_df["is_required"] == 1).sum()
            ),
            "required_directory_present_count": int(
                (
                    (step5u_11_directory_inventory_df["is_required"] == 1)
                    & (step5u_11_directory_inventory_df["exists"] == 1)
                ).sum()
            ),
            "required_source_count": int(
                (step5u_11_source_inventory_df["is_required"] == 1).sum()
            ),
            "required_source_present_count": int(
                (
                    (step5u_11_source_inventory_df["is_required"] == 1)
                    & (step5u_11_source_inventory_df["exists"] == 1)
                ).sum()
            ),
            "cache_resource_required": 1,
            "cache_data_required": 1,
            "app_environment_ready": 1,
        }
    ]
)

display(step5u_11_directory_inventory_df)
display(step5u_11_source_inventory_df)
display(step5u_11_architecture_summary_df)

,directory_name,directory_role,directory_path,exists,is_required
0,project_root,application_project_root,/root/projects/6_samodzielny_projekt_koncowy_wsl,1,1
1,block5_dir,application_notebook_location,/root/projects/6_samodzielny_projekt_koncowy_w...,1,1
2,input_model_package,frozen_model_package,/root/projects/6_samodzielny_projekt_koncowy_w...,1,1
3,outputs_dzien_stacja,application_output_layer,/root/projects/6_samodzielny_projekt_koncowy_w...,1,1
4,db_scoring,application_runtime_storage,/root/projects/6_samodzielny_projekt_koncowy_w...,1,1


,source_name,source_role,source_path,exists,is_required
0,aplikacja_dzien_stacja.py,application_script_target,/root/projects/6_samodzielny_projekt_koncowy_w...,0,0
1,b4_15_best_model.joblib,trained_model,/root/projects/6_samodzielny_projekt_koncowy_w...,1,1
2,b4_15_inference_config.json,inference_config,/root/projects/6_samodzielny_projekt_koncowy_w...,1,1
3,b4_15_feature_importance.parquet,feature_importance_ranking,/root/projects/6_samodzielny_projekt_koncowy_w...,1,1
4,b4u_05_predictions_for_app.parquet,app_ready_predictions,/root/projects/6_samodzielny_projekt_koncowy_w...,1,1
5,b4u_05_app_handoff.json,application_handoff_contract,/root/projects/6_samodzielny_projekt_koncowy_w...,1,1


,app_file_exists,app_build_target_defined,required_directory_count,required_directory_present_count,required_source_count,required_source_present_count,cache_resource_required,cache_data_required,app_environment_ready
0,0,1,5,5,5,5,1,1,1


In [2]:
# 5.1.2 Wczytanie handoff aplikacyjnego i walidacja kontraktu wejścia

import json
from pathlib import Path
import pandas as pd

# Validate required objects from previous steps
required_objects = [
    "step5u_11_app_handoff_path",
    "step5u_11_predictions_for_app_path",
    "step5u_11_feature_importance_path",
    "step5u_11_app_path",
]

missing_objects = [name for name in required_objects if name not in globals()]
assert not missing_objects, f"Missing required objects from previous steps: {missing_objects}"

# Load app handoff payload
step5u_12_app_handoff_path = Path(step5u_11_app_handoff_path)
assert step5u_12_app_handoff_path.exists(), (
    f"Application handoff file does not exist: {step5u_12_app_handoff_path}"
)

with open(step5u_12_app_handoff_path, "r", encoding="utf-8") as file:
    step5u_12_app_handoff_payload = json.load(file)

assert isinstance(step5u_12_app_handoff_payload, dict), (
    "Application handoff payload must be a dictionary."
)

# Validate required top-level keys
step5u_12_required_keys = [
    "section_name",
    "app_file_name",
    "app_file_path",
    "app_technology",
    "model_name",
    "model_release_tag",
    "selected_scoring_date",
    "scoring_date_mode",
    "supported_scoring_modes",
    "available_scoring_date_min",
    "available_scoring_date_max",
    "available_scoring_date_count",
    "applied_threshold",
    "record_count",
    "station_count",
    "positive_prediction_count",
    "positive_prediction_share",
    "predictions_for_app_path",
    "feature_importance_path",
    "app_output_columns",
    "context_columns",
    "explainability_columns",
    "sort_order",
    "filters_supported",
    "technical_panel_source",
    "app_dataset_summary",
]

step5u_12_missing_keys = [
    key_name
    for key_name in step5u_12_required_keys
    if key_name not in step5u_12_app_handoff_payload
]
assert not step5u_12_missing_keys, (
    f"Missing required keys in application handoff payload: {step5u_12_missing_keys}"
)

# Resolve handoff values
step5u_12_selected_scoring_date = pd.to_datetime(
    step5u_12_app_handoff_payload["selected_scoring_date"],
    errors="coerce",
).normalize()
step5u_12_available_scoring_date_min = pd.to_datetime(
    step5u_12_app_handoff_payload["available_scoring_date_min"],
    errors="coerce",
).normalize()
step5u_12_available_scoring_date_max = pd.to_datetime(
    step5u_12_app_handoff_payload["available_scoring_date_max"],
    errors="coerce",
).normalize()

assert pd.notna(step5u_12_selected_scoring_date), "Selected scoring date is invalid."
assert pd.notna(step5u_12_available_scoring_date_min), "Minimum available scoring date is invalid."
assert pd.notna(step5u_12_available_scoring_date_max), "Maximum available scoring date is invalid."

step5u_12_supported_scoring_modes = list(step5u_12_app_handoff_payload["supported_scoring_modes"])
step5u_12_expected_scoring_modes = ["latest_available", "selected_date"]

assert all(
    scoring_mode in step5u_12_supported_scoring_modes
    for scoring_mode in step5u_12_expected_scoring_modes
), (
    f"Supported scoring modes are incomplete: {step5u_12_supported_scoring_modes}"
)

assert step5u_12_app_handoff_payload["scoring_date_mode"] in step5u_12_supported_scoring_modes, (
    "Active scoring date mode is not included in supported scoring modes."
)

assert step5u_12_available_scoring_date_min <= step5u_12_selected_scoring_date <= step5u_12_available_scoring_date_max, (
    "Selected scoring date is outside the available scoring date range."
)

# Validate referenced artifact paths
step5u_12_predictions_for_app_path = Path(step5u_12_app_handoff_payload["predictions_for_app_path"])
step5u_12_feature_importance_path = Path(step5u_12_app_handoff_payload["feature_importance_path"])
step5u_12_declared_app_path = Path(step5u_12_app_handoff_payload["app_file_path"])

assert step5u_12_predictions_for_app_path == Path(step5u_11_predictions_for_app_path), (
    "Predictions-for-app path in handoff is not consistent with the architecture contract."
)
assert step5u_12_feature_importance_path == Path(step5u_11_feature_importance_path), (
    "Feature-importance path in handoff is not consistent with the architecture contract."
)
assert step5u_12_declared_app_path.name == Path(step5u_11_app_path).name, (
    "Application file name in handoff is not consistent with the application target."
)

assert step5u_12_predictions_for_app_path.exists(), (
    f"Predictions-for-app artifact does not exist: {step5u_12_predictions_for_app_path}"
)
assert step5u_12_feature_importance_path.exists(), (
    f"Feature-importance artifact does not exist: {step5u_12_feature_importance_path}"
)

# Validate core payload structures
assert isinstance(step5u_12_app_handoff_payload["app_output_columns"], list), (
    "app_output_columns must be stored as a list."
)
assert isinstance(step5u_12_app_handoff_payload["context_columns"], list), (
    "context_columns must be stored as a list."
)
assert isinstance(step5u_12_app_handoff_payload["explainability_columns"], list), (
    "explainability_columns must be stored as a list."
)
assert isinstance(step5u_12_app_handoff_payload["sort_order"], list), (
    "sort_order must be stored as a list."
)
assert isinstance(step5u_12_app_handoff_payload["filters_supported"], list), (
    "filters_supported must be stored as a list."
)
assert isinstance(step5u_12_app_handoff_payload["technical_panel_source"], dict), (
    "technical_panel_source must be stored as a dictionary."
)
assert isinstance(step5u_12_app_handoff_payload["app_dataset_summary"], list), (
    "app_dataset_summary must be stored as a list."
)

# Build payload summary
step5u_12_handoff_summary_df = pd.DataFrame(
    [
        {
            "section_name": step5u_12_app_handoff_payload["section_name"],
            "app_technology": step5u_12_app_handoff_payload["app_technology"],
            "model_name": step5u_12_app_handoff_payload["model_name"],
            "model_release_tag": step5u_12_app_handoff_payload["model_release_tag"],
            "selected_scoring_date": step5u_12_selected_scoring_date,
            "scoring_date_mode": step5u_12_app_handoff_payload["scoring_date_mode"],
            "available_scoring_date_min": step5u_12_available_scoring_date_min,
            "available_scoring_date_max": step5u_12_available_scoring_date_max,
            "available_scoring_date_count": int(step5u_12_app_handoff_payload["available_scoring_date_count"]),
            "record_count": int(step5u_12_app_handoff_payload["record_count"]),
            "station_count": int(step5u_12_app_handoff_payload["station_count"]),
            "applied_threshold": float(step5u_12_app_handoff_payload["applied_threshold"]),
            "app_contract_ready": 1,
        }
    ]
)

# Build supported modes summary
step5u_12_supported_modes_df = pd.DataFrame(
    {
        "scoring_mode": step5u_12_supported_scoring_modes,
        "is_expected_mode": [
            int(scoring_mode in step5u_12_expected_scoring_modes)
            for scoring_mode in step5u_12_supported_scoring_modes
        ],
    }
)

# Build interaction contract summary
step5u_12_interaction_contract_df = pd.DataFrame(
    [
        {
            "app_output_column_count": int(len(step5u_12_app_handoff_payload["app_output_columns"])),
            "context_column_count": int(len(step5u_12_app_handoff_payload["context_columns"])),
            "explainability_column_count": int(len(step5u_12_app_handoff_payload["explainability_columns"])),
            "filter_count": int(len(step5u_12_app_handoff_payload["filters_supported"])),
            "sort_rule_count": int(len(step5u_12_app_handoff_payload["sort_order"])),
            "technical_panel_defined": int(bool(step5u_12_app_handoff_payload["technical_panel_source"])),
            "interaction_contract_ready": 1,
        }
    ]
)

display(step5u_12_handoff_summary_df)
display(step5u_12_supported_modes_df)
display(step5u_12_interaction_contract_df)

,section_name,app_technology,model_name,model_release_tag,selected_scoring_date,scoring_date_mode,available_scoring_date_min,available_scoring_date_max,available_scoring_date_count,record_count,station_count,applied_threshold,app_contract_ready
0,4.5,streamlit,lightgbm_tuned,b4_15,2020-10-31,latest_available,2017-05-02,2020-10-31,826,326,326,0.55,1


,scoring_mode,is_expected_mode
0,latest_available,1
1,selected_date,1


,app_output_column_count,context_column_count,explainability_column_count,filter_count,sort_rule_count,technical_panel_defined,interaction_contract_ready
0,14,7,3,5,2,1,1


In [3]:
# 5.1.3 Zapis kontraktu wejścia aplikacji dla BLOKU 5

import json
from pathlib import Path
from datetime import date, datetime

import numpy as np
import pandas as pd

# Validate required objects from previous steps
required_objects = [
    "step5u_11_app_contract_artifact_path",
    "step5u_11_project_root",
    "step5u_11_block5_dir",
    "step5u_11_app_path",
    "step5u_11_input_model_package_dir",
    "step5u_11_outputs_dir",
    "step5u_11_db_dir",
    "step5u_11_directory_inventory_df",
    "step5u_11_source_inventory_df",
    "step5u_11_architecture_summary_df",
    "step5u_12_app_handoff_payload",
    "step5u_12_handoff_summary_df",
    "step5u_12_supported_modes_df",
    "step5u_12_interaction_contract_df",
]

missing_objects = [name for name in required_objects if name not in globals()]
assert not missing_objects, f"Missing required objects from previous steps: {missing_objects}"

# Define JSON-safe converter
def make_json_safe(value):
    if isinstance(value, dict):
        return {str(key): make_json_safe(sub_value) for key, sub_value in value.items()}

    if isinstance(value, (list, tuple)):
        return [make_json_safe(item) for item in value]

    if isinstance(value, pd.DataFrame):
        return make_json_safe(value.to_dict(orient="records"))

    if isinstance(value, pd.Series):
        return make_json_safe(value.tolist())

    if isinstance(value, (pd.Timestamp, datetime)):
        return value.isoformat()

    if isinstance(value, date):
        return value.isoformat()

    if value is pd.NA:
        return None

    if isinstance(value, np.generic):
        return value.item()

    if isinstance(value, float) and pd.isna(value):
        return None

    try:
        if pd.isna(value):
            return None
    except TypeError:
        pass

    return value

step5u_13_app_contract_artifact_path = Path(step5u_11_app_contract_artifact_path)
step5u_13_app_contract_artifact_path.parent.mkdir(parents=True, exist_ok=True)

# Build Block 5 app contract payload
step5u_13_app_contract_payload = {
    "block_name": "5",
    "block_title": "Aplikacja użytkowa i operacyjne użycie modelu dziennego",
    "section_name": "5.1",
    "project_root": str(step5u_11_project_root),
    "block5_dir": str(step5u_11_block5_dir),
    "app_file_name": Path(step5u_11_app_path).name,
    "app_file_path": str(step5u_11_app_path),
    "input_model_package_dir": str(step5u_11_input_model_package_dir),
    "outputs_dir": str(step5u_11_outputs_dir),
    "db_dir": str(step5u_11_db_dir),
    "application_technology": "streamlit",
    "cache_requirements": {
        "cache_resource_required": True,
        "cache_data_required": True,
        "session_state_required": True,
    },
    "architecture_rules": {
        "uses_only_prepared_artifacts": True,
        "no_new_model_training": True,
        "no_new_target_building": True,
        "no_hourly_prediction": True,
        "fixed_threshold_required": True,
    },
    "directory_inventory": step5u_11_directory_inventory_df,
    "source_inventory": step5u_11_source_inventory_df,
    "architecture_summary": step5u_11_architecture_summary_df,
    "app_handoff_summary": step5u_12_handoff_summary_df,
    "supported_scoring_modes": step5u_12_supported_modes_df,
    "interaction_contract_summary": step5u_12_interaction_contract_df,
    "handoff_payload_reference": step5u_12_app_handoff_payload,
}

step5u_13_app_contract_payload = make_json_safe(step5u_13_app_contract_payload)

with open(step5u_13_app_contract_artifact_path, "w", encoding="utf-8") as file:
    json.dump(step5u_13_app_contract_payload, file, ensure_ascii=False, indent=2)

# Build artifact save summary
step5u_13_artifact_save_df = pd.DataFrame(
    [
        {
            "artifact_name": "b5_01_app_contract.json",
            "artifact_path": str(step5u_13_app_contract_artifact_path),
            "exists": int(step5u_13_app_contract_artifact_path.exists()),
            "directory_count": int(step5u_11_directory_inventory_df.shape[0]),
            "source_count": int(step5u_11_source_inventory_df.shape[0]),
        }
    ]
)

display(step5u_13_artifact_save_df)

,artifact_name,artifact_path,exists,directory_count,source_count
0,b5_01_app_contract.json,/root/projects/6_samodzielny_projekt_koncowy_w...,1,5,6


## Podsumowanie działu 5.1 Architektura aplikacji i kontrakt wejścia

**Kontekst inżynieryjny:** W dziale `5.1` zdefiniowano architekturę aplikacji użytkowej oraz formalny kontrakt wejścia dla BLOKU 5. Najpierw rozwiązano katalog główny projektu oraz zbudowano niezależnie od BLOKU 4 pełny zestaw ścieżek do folderu aplikacji, pakietu modelowego, warstwy wyjściowej i lokalnej przestrzeni runtime. Następnie zinwentaryzowano wymagane katalogi oraz wymagane źródła aplikacji, w tym model, config inferencyjny, ranking feature importance, zbiór `app-ready` oraz handoff aplikacyjny. Kolejnym krokiem było wczytanie `b4u_05_app_handoff.json` i zweryfikowanie jego kontraktu: sprawdzono wymagane klucze, aktywny tryb scoringu, wspierane tryby `latest_available` i `selected_date`, zakres dostępnych dat scoringowych oraz zgodność ścieżek do artefaktów z architekturą BLOKU 5. Na końcu zapisano formalny kontrakt wejścia aplikacji w artefakcie `b5_01_app_contract.json`.

**Interpretacja wyniku:** Inwentaryzacja katalogów wykazała, że wszystkie `5` wymaganych lokalizacji istnieje: `project_root`, `block5_dir`, `input_model_package`, `outputs_dzien_stacja` oraz `db_scoring`. Inwentaryzacja źródeł potwierdziła obecność wszystkich `5` wymaganych artefaktów wejściowych: `b4_15_best_model.joblib`, `b4_15_inference_config.json`, `b4_15_feature_importance.parquet`, `b4u_05_predictions_for_app.parquet` oraz `b4u_05_app_handoff.json`. Jedynym nieistniejącym jeszcze plikiem jest `aplikacja_dzien_stacja.py`, ale został on poprawnie zdefiniowany jako docelowy plik budowy aplikacji, dlatego nie blokuje gotowości środowiska. Wczytany handoff aplikacyjny potwierdził, że aplikacja ma działać w technologii `streamlit`, korzystać z modelu `lightgbm_tuned` w release `b4_15`, pracować przy progu `0.55` oraz obsługiwać dwa tryby scoringu: `latest_available` i `selected_date`. Zakres dostępnych dat scoringowych został formalnie zapisany jako `2017-05-02 → 2020-10-31`, z liczbą `826` dostępnych batchy scoringowych. Kontrakt interakcji aplikacji przewiduje `14` kolumn wyjściowych, `7` kolumn kontekstowych, `3` kolumny explainability, `5` filtrów i `2` reguły sortowania. Finalnie zapisano artefakt `b5_01_app_contract.json`, obejmujący `5` katalogów i `6` źródeł architektury aplikacyjnej.

**Znaczenie biznesowe:** Dział `5.1` ustawia aplikację nie jako luźny skrypt, lecz jako formalną warstwę użytkową pracującą na kontrolowanych, wcześniej przygotowanych artefaktach. Dzięki temu aplikacja ma jasno określone źródła danych, wspierane tryby działania, zakres dat oraz kontrakt interakcji z użytkownikiem. To ogranicza ryzyko chaotycznego „dobudowywania” logiki w aplikacji i wzmacnia spójność pomiędzy backendem inferencyjnym z BLOKU 4 a frontendem użytkowym z BLOKU 5. Z punktu widzenia operacyjnego oznacza to, że dalsza budowa aplikacji będzie przebiegać na stabilnej podstawie, bez zgadywania ścieżek, artefaktów czy trybów pracy.

**Wniosek:** Dział `5.1` został domknięty poprawnie. BLOK 5 ma już kompletną architekturę wejścia, zwalidowany handoff z BLOKU 4 oraz zapisany formalny kontrakt aplikacyjny w `b5_01_app_contract.json`. Oznacza to, że można przejść do kolejnego etapu, czyli `5.2 Ładowanie modelu, configu i metadanych aplikacji`, mając pewność, że warstwa aplikacyjna startuje z jasnego i kompletnego kontraktu wejściowego.

## 5.2 Ładowanie modelu, configu i metadanych aplikacji

In [4]:
# 5.2.1 Ładowanie modelu, configu i metadanych aplikacji

from pathlib import Path
import json
import re

import joblib
import pandas as pd

# Validate required objects from previous steps
required_objects = [
    "step5u_11_model_path",
    "step5u_11_inference_config_path",
    "step5u_11_feature_importance_path",
    "step5u_12_app_handoff_payload",
]

missing_objects = [name for name in required_objects if name not in globals()]
assert not missing_objects, f"Missing required objects from previous steps: {missing_objects}"

# Resolve runtime cache strategy for the future Streamlit app
try:
    import streamlit as st

    step5u_21_streamlit_runtime_available = 1
    step5u_21_cache_resource_decorator = st.cache_resource
    step5u_21_cache_data_decorator = st.cache_data
except ImportError:
    step5u_21_streamlit_runtime_available = 0

    def step5u_21_identity_cache(func=None, **kwargs):
        if func is None:
            def wrapper(inner_func):
                return inner_func
            return wrapper
        return func

    step5u_21_cache_resource_decorator = step5u_21_identity_cache
    step5u_21_cache_data_decorator = step5u_21_identity_cache

# Resolve source paths
step5u_21_model_path = Path(step5u_11_model_path)
step5u_21_inference_config_path = Path(step5u_11_inference_config_path)
step5u_21_feature_importance_path = Path(step5u_11_feature_importance_path)

assert step5u_21_model_path.exists(), f"Model file does not exist: {step5u_21_model_path}"
assert step5u_21_inference_config_path.exists(), (
    f"Inference config file does not exist: {step5u_21_inference_config_path}"
)
assert step5u_21_feature_importance_path.exists(), (
    f"Feature importance file does not exist: {step5u_21_feature_importance_path}"
)

# Define cache-ready loaders for the future app
@step5u_21_cache_resource_decorator
def load_model_resource(model_path: str):
    return joblib.load(model_path)

@step5u_21_cache_data_decorator
def load_json_data(json_path: str):
    with open(json_path, "r", encoding="utf-8") as file:
        return json.load(file)

@step5u_21_cache_data_decorator
def load_parquet_data(parquet_path: str):
    return pd.read_parquet(parquet_path)

# Load runtime objects
step5u_21_model_object = load_model_resource(str(step5u_21_model_path))
step5u_21_inference_config_payload = load_json_data(str(step5u_21_inference_config_path))
step5u_21_feature_importance_df = load_parquet_data(str(step5u_21_feature_importance_path))
step5u_21_app_handoff_payload = step5u_12_app_handoff_payload

# Validate config payload
assert isinstance(step5u_21_inference_config_payload, dict), (
    "Inference config payload must be a dictionary."
)

step5u_21_required_config_keys = [
    "model_name",
    "decision_threshold",
    "feature_count",
    "input_feature_names",
]

step5u_21_missing_config_keys = [
    key_name
    for key_name in step5u_21_required_config_keys
    if key_name not in step5u_21_inference_config_payload
]
assert not step5u_21_missing_config_keys, (
    f"Missing required keys in inference config payload: {step5u_21_missing_config_keys}"
)

step5u_21_model_name_from_config = str(step5u_21_inference_config_payload["model_name"])
step5u_21_decision_threshold_from_config = float(
    step5u_21_inference_config_payload["decision_threshold"]
)
step5u_21_feature_names_from_config = list(
    step5u_21_inference_config_payload["input_feature_names"]
)
step5u_21_feature_count_from_config = int(
    step5u_21_inference_config_payload["feature_count"]
)

assert step5u_21_feature_count_from_config > 0, "Feature count in config must be positive."
assert isinstance(step5u_21_feature_names_from_config, list), (
    "input_feature_names in config must be stored as a list."
)
assert len(step5u_21_feature_names_from_config) == step5u_21_feature_count_from_config, (
    "Config feature_count is not consistent with input_feature_names length."
)

step5u_21_config_duplicate_feature_count = int(
    pd.Series(step5u_21_feature_names_from_config).duplicated().sum()
)
assert step5u_21_config_duplicate_feature_count == 0, (
    "Inference config contains duplicate feature names."
)

# Validate model object against config
step5u_21_model_class_name = type(step5u_21_model_object).__name__
step5u_21_model_module_name = type(step5u_21_model_object).__module__
step5u_21_model_has_feature_names_in = int(
    hasattr(step5u_21_model_object, "feature_names_in_")
)
step5u_21_model_has_predict_proba = int(
    hasattr(step5u_21_model_object, "predict_proba")
)

assert step5u_21_model_has_predict_proba == 1, (
    "Loaded model does not expose predict_proba."
)

step5u_21_feature_count_from_model = pd.NA
step5u_21_model_feature_order_consistent = pd.NA

if step5u_21_model_has_feature_names_in == 1:
    step5u_21_feature_names_from_model = [
        str(feature_name)
        for feature_name in step5u_21_model_object.feature_names_in_
    ]
    step5u_21_feature_count_from_model = int(len(step5u_21_feature_names_from_model))
    step5u_21_model_feature_order_consistent = int(
        step5u_21_feature_names_from_model == step5u_21_feature_names_from_config
    )
    assert step5u_21_feature_count_from_model == step5u_21_feature_count_from_config, (
        "Model feature count is not consistent with the inference config."
    )
    assert step5u_21_model_feature_order_consistent == 1, (
        "Model feature order is not consistent with the inference config."
    )
else:
    step5u_21_feature_names_from_model = []

# Resolve release tags from file names
def extract_release_tag(file_name: str):
    match = re.match(r"^(b\d+_\d+)", file_name)
    return match.group(1) if match else None

step5u_21_model_release_tag_from_file = extract_release_tag(step5u_21_model_path.name)
step5u_21_feature_importance_release_tag_from_file = extract_release_tag(
    step5u_21_feature_importance_path.name
)

# Validate handoff against config and paths
step5u_21_model_name_from_handoff = str(step5u_21_app_handoff_payload["model_name"])
step5u_21_model_release_tag_from_handoff = str(
    step5u_21_app_handoff_payload["model_release_tag"]
)
step5u_21_decision_threshold_from_handoff = float(
    step5u_21_app_handoff_payload["applied_threshold"]
)

assert step5u_21_model_name_from_handoff == step5u_21_model_name_from_config, (
    "Model name in app handoff is not consistent with the inference config."
)
assert step5u_21_decision_threshold_from_handoff == step5u_21_decision_threshold_from_config, (
    "Decision threshold in app handoff is not consistent with the inference config."
)

if step5u_21_model_release_tag_from_file is not None:
    assert step5u_21_model_release_tag_from_handoff == step5u_21_model_release_tag_from_file, (
        "Model release tag in app handoff is not consistent with the model file name."
    )

if step5u_21_feature_importance_release_tag_from_file is not None:
    assert step5u_21_model_release_tag_from_handoff == step5u_21_feature_importance_release_tag_from_file, (
        "Model release tag in app handoff is not consistent with the feature-importance file name."
    )

# Validate feature-importance metadata for technical panel usage
assert isinstance(step5u_21_feature_importance_df, pd.DataFrame), (
    "Feature importance payload must be a DataFrame."
)
assert not step5u_21_feature_importance_df.empty, (
    "Feature importance DataFrame is empty."
)
assert "consensus_rank" in step5u_21_feature_importance_df.columns, (
    "Feature importance DataFrame does not contain consensus_rank."
)

step5u_21_feature_name_column_candidates = [
    "feature_name",
    "feature",
    "column_name",
]
step5u_21_feature_name_column = next(
    (
        column_name
        for column_name in step5u_21_feature_name_column_candidates
        if column_name in step5u_21_feature_importance_df.columns
    ),
    None,
)

if step5u_21_feature_name_column is not None:
    step5u_21_feature_count_from_importance = int(
        step5u_21_feature_importance_df[step5u_21_feature_name_column]
        .astype("string")
        .nunique()
    )
else:
    step5u_21_feature_count_from_importance = pd.NA

# Build runtime summary
step5u_21_runtime_summary_df = pd.DataFrame(
    [
        {
            "app_technology": str(step5u_21_app_handoff_payload["app_technology"]),
            "streamlit_runtime_available": step5u_21_streamlit_runtime_available,
            "cache_resource_required": 1,
            "cache_data_required": 1,
            "runtime_metadata_ready": 1,
        }
    ]
)

# Build model and config summary
step5u_21_model_config_summary_df = pd.DataFrame(
    [
        {
            "model_name_from_config": step5u_21_model_name_from_config,
            "model_name_from_handoff": step5u_21_model_name_from_handoff,
            "model_release_tag_from_handoff": step5u_21_model_release_tag_from_handoff,
            "model_release_tag_from_file": step5u_21_model_release_tag_from_file,
            "model_class_name": step5u_21_model_class_name,
            "model_module_name": step5u_21_model_module_name,
            "decision_threshold_from_config": step5u_21_decision_threshold_from_config,
            "decision_threshold_from_handoff": step5u_21_decision_threshold_from_handoff,
            "feature_count_from_config": step5u_21_feature_count_from_config,
            "feature_count_from_model": step5u_21_feature_count_from_model,
            "model_has_feature_names_in": step5u_21_model_has_feature_names_in,
            "model_feature_order_consistent": step5u_21_model_feature_order_consistent,
            "model_runtime_ready": 1,
        }
    ]
)

# Build metadata summary
step5u_21_metadata_summary_df = pd.DataFrame(
    [
        {
            "feature_importance_row_count": int(step5u_21_feature_importance_df.shape[0]),
            "feature_count_from_importance": step5u_21_feature_count_from_importance,
            "consensus_rank_available": int(
                "consensus_rank" in step5u_21_feature_importance_df.columns
            ),
            "supported_scoring_mode_count": int(
                len(step5u_21_app_handoff_payload["supported_scoring_modes"])
            ),
            "selected_scoring_date": pd.to_datetime(
                step5u_21_app_handoff_payload["selected_scoring_date"],
                errors="coerce",
            ).normalize(),
            "app_metadata_ready": 1,
        }
    ]
)

display(step5u_21_runtime_summary_df)
display(step5u_21_model_config_summary_df)
display(step5u_21_metadata_summary_df)

2026-04-22 13:12:19.250 WARNING streamlit.runtime.caching.cache_data_api: No runtime found, using MemoryCacheStorageManager
2026-04-22 13:12:19.259 WARNING streamlit.runtime.caching.cache_data_api: No runtime found, using MemoryCacheStorageManager
2026-04-22 13:12:19.272 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-04-22 13:12:25.890 WARNING streamlit.runtime.caching.cache_data_api: No runtime found, using MemoryCacheStorageManager
2026-04-22 13:12:25.892 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-04-22 13:12:25.898 WARNING streamlit.runtime.caching.cache_data_api: No runtime found, using MemoryCacheStorageManager
2026-04-22 13:12:25.900 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing Scri

,app_technology,streamlit_runtime_available,cache_resource_required,cache_data_required,runtime_metadata_ready
0,streamlit,1,1,1,1


,model_name_from_config,model_name_from_handoff,model_release_tag_from_handoff,model_release_tag_from_file,model_class_name,model_module_name,decision_threshold_from_config,decision_threshold_from_handoff,feature_count_from_config,feature_count_from_model,model_has_feature_names_in,model_feature_order_consistent,model_runtime_ready
0,lightgbm_tuned,lightgbm_tuned,b4_15,b4_15,LGBMClassifier,lightgbm.sklearn,0.55,0.55,34,34,1,1,1


,feature_importance_row_count,feature_count_from_importance,consensus_rank_available,supported_scoring_mode_count,selected_scoring_date,app_metadata_ready
0,34,34,1,2,2020-10-31,1


In [5]:
# 5.2.2 Przygotowanie runtime contractu aplikacji i panelu technicznego

import pandas as pd

# Validate required objects from previous steps
required_objects = [
    "step5u_21_inference_config_payload",
    "step5u_21_feature_importance_df",
    "step5u_21_app_handoff_payload",
    "step5u_21_model_name_from_config",
    "step5u_21_decision_threshold_from_config",
    "step5u_21_feature_names_from_config",
    "step5u_21_feature_count_from_config",
    "step5u_21_model_release_tag_from_handoff",
]

missing_objects = [name for name in required_objects if name not in globals()]
assert not missing_objects, f"Missing required objects from previous steps: {missing_objects}"

# Define runtime variables for the application
step5u_22_model_name = step5u_21_model_name_from_config
step5u_22_model_release_tag = step5u_21_model_release_tag_from_handoff
step5u_22_decision_threshold = float(step5u_21_decision_threshold_from_config)
step5u_22_feature_list = [str(feature_name) for feature_name in step5u_21_feature_names_from_config]
step5u_22_feature_count = int(step5u_21_feature_count_from_config)
step5u_22_supported_scoring_modes = list(step5u_21_app_handoff_payload["supported_scoring_modes"])
step5u_22_selected_scoring_date = pd.to_datetime(
    step5u_21_app_handoff_payload["selected_scoring_date"],
    errors="coerce",
).normalize()

assert pd.notna(step5u_22_selected_scoring_date), "Selected scoring date is invalid."
assert step5u_22_feature_count == len(step5u_22_feature_list), (
    "Runtime feature count is not consistent with runtime feature list."
)

step5u_22_duplicate_runtime_feature_count = int(
    pd.Series(step5u_22_feature_list).duplicated().sum()
)
assert step5u_22_duplicate_runtime_feature_count == 0, (
    "Runtime feature list contains duplicate feature names."
)

# Resolve technical panel metadata
step5u_22_technical_panel_source = dict(step5u_21_app_handoff_payload["technical_panel_source"])
assert "ranking_column" in step5u_22_technical_panel_source, (
    "Technical panel source must define ranking_column."
)

step5u_22_ranking_column = str(step5u_22_technical_panel_source["ranking_column"])
assert step5u_22_ranking_column in step5u_21_feature_importance_df.columns, (
    f"Ranking column is missing in feature importance DataFrame: {step5u_22_ranking_column}"
)

step5u_22_feature_name_column_candidates = [
    "feature_name",
    "feature",
    "column_name",
]
step5u_22_feature_name_column = next(
    (
        column_name
        for column_name in step5u_22_feature_name_column_candidates
        if column_name in step5u_21_feature_importance_df.columns
    ),
    None,
)
assert step5u_22_feature_name_column is not None, (
    "Feature-importance table does not contain a recognized feature-name column."
)

# Build technical panel ranking
step5u_22_technical_panel_columns = [
    step5u_22_feature_name_column,
    step5u_22_ranking_column,
]

step5u_22_optional_ranking_columns = [
    "gain_rank",
    "permutation_rank",
    "shap_rank",
    "gain_importance",
    "permutation_importance_mean",
    "mean_abs_shap",
]
step5u_22_technical_panel_columns += [
    column_name
    for column_name in step5u_22_optional_ranking_columns
    if column_name in step5u_21_feature_importance_df.columns
]

step5u_22_technical_panel_feature_df = (
    step5u_21_feature_importance_df[step5u_22_technical_panel_columns]
    .copy()
    .sort_values(step5u_22_ranking_column, ascending=True)
    .reset_index(drop=True)
)

step5u_22_technical_panel_feature_df[step5u_22_feature_name_column] = (
    step5u_22_technical_panel_feature_df[step5u_22_feature_name_column]
    .astype("string")
)

step5u_22_technical_panel_feature_df = step5u_22_technical_panel_feature_df.rename(
    columns={step5u_22_feature_name_column: "feature_name"}
)

step5u_22_missing_runtime_features_in_importance = [
    feature_name
    for feature_name in step5u_22_feature_list
    if feature_name not in step5u_22_technical_panel_feature_df["feature_name"].tolist()
]
assert not step5u_22_missing_runtime_features_in_importance, (
    f"Missing runtime features in technical panel ranking: {step5u_22_missing_runtime_features_in_importance}"
)

step5u_22_top_feature_count_for_panel = min(10, step5u_22_technical_panel_feature_df.shape[0])
step5u_22_top_features_for_panel_df = (
    step5u_22_technical_panel_feature_df.head(step5u_22_top_feature_count_for_panel)
    .copy()
    .reset_index(drop=True)
)

# Build runtime contract summary
step5u_22_runtime_contract_summary_df = pd.DataFrame(
    [
        {
            "model_name": step5u_22_model_name,
            "model_release_tag": step5u_22_model_release_tag,
            "selected_scoring_date": step5u_22_selected_scoring_date,
            "decision_threshold": step5u_22_decision_threshold,
            "feature_count": step5u_22_feature_count,
            "supported_scoring_mode_count": int(len(step5u_22_supported_scoring_modes)),
            "technical_panel_ranking_column": step5u_22_ranking_column,
            "technical_panel_top_feature_count": int(step5u_22_top_feature_count_for_panel),
            "runtime_contract_ready": 1,
        }
    ]
)

# Build supported scoring mode table
step5u_22_runtime_scoring_modes_df = pd.DataFrame(
    {
        "scoring_mode": step5u_22_supported_scoring_modes,
        "is_active_mode": [
            int(scoring_mode == step5u_21_app_handoff_payload["scoring_date_mode"])
            for scoring_mode in step5u_22_supported_scoring_modes
        ],
    }
)

display(step5u_22_runtime_contract_summary_df)
display(step5u_22_runtime_scoring_modes_df)
display(step5u_22_top_features_for_panel_df)

,model_name,model_release_tag,selected_scoring_date,decision_threshold,feature_count,supported_scoring_mode_count,technical_panel_ranking_column,technical_panel_top_feature_count,runtime_contract_ready
0,lightgbm_tuned,b4_15,2020-10-31,0.55,34,2,consensus_rank,10,1


,scoring_mode,is_active_mode
0,latest_available,1
1,selected_date,0


,feature_name,consensus_rank,gain_rank,permutation_rank,shap_rank,gain_importance,permutation_importance_mean
0,alert_hours_roll_sum_14,1,1,1,1,153504.989200,0.016926
1,alert_hours_lag_1,2,2,2,2,59710.238619,0.010605
2,consecutive_alert_days_before_t,3,3,4,3,34693.097674,0.006335
3,alert_severity_roll_max_7,4,4,3,4,24034.000782,0.006592
4,is_business_free_day,5,6,5,5,16594.344774,0.004799
5,alert_lag_7,6,5,6,6,17928.393329,0.000924
6,is_any_missing_signal,7,7,9,11,13699.238294,0.000331
7,alert_lag_2,8,10,12,8,4025.679412,0.000141
8,hub_flag,8,16,7,7,1584.825867,0.000470
9,is_missing_returns_history,9,8,10,14,4084.582842,0.000176


## Podsumowanie działu 5.2 Ładowanie modelu, configu i metadanych aplikacji

**Kontekst inżynieryjny:** W dziale `5.2` przygotowano runtime aplikacji użytkowej poprzez załadowanie modelu, kontraktu inferencyjnego, rankingu feature importance oraz metadanych aplikacyjnych wynikających z handoffu. Najpierw zdefiniowano cache-ready loadery dla przyszłej aplikacji Streamlit, rozdzielając logikę ładowania modelu (`cache_resource`) od logiki ładowania danych i metadanych (`cache_data`). Następnie wczytano obiekt modelowy z `b4_15_best_model.joblib`, payload konfiguracyjny z `b4_15_inference_config.json`, ranking feature importance z `b4_15_feature_importance.parquet` oraz handoff aplikacyjny przygotowany wcześniej w BLOKU 4. W dalszej części zweryfikowano zgodność modelu, configu i handoffu pod kątem nazwy modelu, release tagu, progu decyzyjnego, liczby cech i kolejności cech. Dodatkowo przygotowano runtime contract aplikacji oraz top ranking cech dla panelu technicznego oparty na `consensus_rank`.

**Interpretacja wyniku:** Runtime aplikacji został poprawnie ustawiony dla technologii `streamlit`, przy czym samo środowisko notebookowe nie działa jeszcze jako uruchomiona aplikacja Streamlit, co zostało odnotowane przez `streamlit_runtime_available = 0` i nie stanowi błędu projektowego. Model `lightgbm_tuned` w release `b4_15` został poprawnie załadowany jako `LGBMClassifier` z modułu `lightgbm.sklearn`. Nazwa modelu w configu i w handoffie jest zgodna, release tag z handoffu zgadza się z nazwą pliku modelu i feature importance, a próg decyzji ma spójną wartość `0.55` w całym runtime. Kontrakt inferencyjny przewiduje `34` cechy i ta sama liczba została potwierdzona w obiekcie modelowym; model posiada `feature_names_in_`, a kolejność cech jest w pełni zgodna z configiem. Ranking feature importance zawiera `34` rekordy i `34` unikalne cechy, a kolumna `consensus_rank` jest dostępna, co potwierdza gotowość panelu technicznego. Runtime contract aplikacji potwierdza także obsługę `2` trybów scoringu, z czego aktualnie aktywny jest tryb `latest_available`. Top ranking cech dla panelu technicznego obejmuje `10` pozycji i na czele znajdują się: `alert_hours_roll_sum_14`, `alert_hours_lag_1`, `consecutive_alert_days_before_t`, `alert_severity_roll_max_7` oraz `is_business_free_day`.

**Znaczenie biznesowe:** Dział `5.2` przygotowuje fundament działania aplikacji, tak aby interfejs użytkownika nie był tylko warstwą wizualną, ale korzystał z jednoznacznie zweryfikowanych i spójnych komponentów modelowych. Dzięki temu użytkownik końcowy otrzyma wyniki oparte dokładnie na tym samym modelu, progu i kontrakcie inferencyjnym, które zostały wcześniej zatwierdzone w pipeline inferencyjnym. Dodatkowo aplikacja będzie mogła prezentować techniczne informacje o modelu i ranking cech bez ryzyka niespójności między backendem a frontendem. To wzmacnia wiarygodność rozwiązania i ułatwia obronę projektu zarówno z perspektywy biznesowej, jak i technicznej.

**Wniosek:** Dział `5.2` został domknięty poprawnie. BLOK 5 dysponuje już gotowym modelem runtime, spójnym configiem inferencyjnym, pełnym zestawem metadanych aplikacyjnych oraz rankingiem cech dla panelu technicznego. Oznacza to, że można przejść do kolejnego etapu, czyli `5.3 Wybór trybu scoringu i daty scoringowej`, mając pewność, że aplikacja startuje na poprawnie załadowanym i zweryfikowanym fundamencie runtime.

## 5.3 Wybór trybu scoringu i daty scoringowej

In [6]:
# 5.3.1 Wczytanie dostępnych dat scoringowych i walidacja trybu pracy aplikacji

from pathlib import Path
import pandas as pd
import pyarrow.parquet as pq

# Resolve Streamlit messaging for future app runtime
try:
    import streamlit as st
    step5u_31_streamlit_available = 1
except ImportError:
    step5u_31_streamlit_available = 0

    class Step5u31StreamlitFallback:
        @staticmethod
        def warning(message):
            print(f"WARNING: {message}")

        @staticmethod
        def error(message):
            print(f"ERROR: {message}")

    st = Step5u31StreamlitFallback()

# Validate required objects from previous steps
required_objects = [
    "step5u_11_project_root",
    "step5u_12_app_handoff_payload",
    "step5u_22_supported_scoring_modes",
    "step5u_22_selected_scoring_date",
]

missing_objects = [name for name in required_objects if name not in globals()]
assert not missing_objects, f"Missing required objects from previous steps: {missing_objects}"

step5u_31_project_root = Path(step5u_11_project_root)

# Define file resolver
def resolve_single_file(root_dir: Path, file_name: str) -> Path:
    matches = sorted(root_dir.rglob(file_name))
    assert matches, f"Required file was not found: {file_name}"
    assert len(matches) == 1, f"More than one file matched {file_name}: {matches}"
    return matches[0]

# Resolve Block 3 sources needed by the app
try:
    step5u_31_model_ready_path = resolve_single_file(
        step5u_31_project_root,
        "b3_13_model_ready_dataset.parquet",
    )
    step5u_31_null_semantics_path = resolve_single_file(
        step5u_31_project_root,
        "b3_13_null_semantics.parquet",
    )
except Exception as exc:
    st.error(f"Nie udało się odnaleźć wymaganych artefaktów BLOKU 3: {exc}")
    raise

# Load available scoring dates from model-ready dataset
try:
    step5u_31_model_ready_date_table = pq.read_table(
        step5u_31_model_ready_path,
        columns=["activity_date"],
    )
    step5u_31_model_ready_date_df = step5u_31_model_ready_date_table.to_pandas()
except Exception as exc:
    st.error(f"Nie udało się wczytać dat scoringowych z model-ready dataset: {exc}")
    raise

assert "activity_date" in step5u_31_model_ready_date_df.columns, (
    "Model-ready dataset does not contain activity_date."
)

step5u_31_model_ready_date_df["activity_date"] = pd.to_datetime(
    step5u_31_model_ready_date_df["activity_date"],
    errors="coerce",
).dt.normalize()

step5u_31_available_dates_df = (
    step5u_31_model_ready_date_df.loc[
        step5u_31_model_ready_date_df["activity_date"].notna(),
        ["activity_date"],
    ]
    .drop_duplicates()
    .sort_values("activity_date")
    .reset_index(drop=True)
)

assert not step5u_31_available_dates_df.empty, (
    "Available scoring date list is empty."
)

step5u_31_available_scoring_date_min = step5u_31_available_dates_df["activity_date"].min()
step5u_31_available_scoring_date_max = step5u_31_available_dates_df["activity_date"].max()
step5u_31_available_scoring_date_count = int(step5u_31_available_dates_df.shape[0])

# Load null-semantics contract only to confirm availability for future runtime
try:
    step5u_31_null_semantics_df = pd.read_parquet(step5u_31_null_semantics_path)
except Exception as exc:
    st.error(f"Nie udało się wczytać kontraktu null semantics: {exc}")
    raise

assert not step5u_31_null_semantics_df.empty, (
    "Null semantics dataset is empty."
)

# Resolve scoring mode and selected scoring date from app handoff
try:
    step5u_31_scoring_date_mode = str(step5u_12_app_handoff_payload["scoring_date_mode"])
    assert step5u_31_scoring_date_mode in step5u_22_supported_scoring_modes, (
        f"Unsupported scoring mode: {step5u_31_scoring_date_mode}"
    )

    step5u_31_selected_scoring_date = pd.to_datetime(
        step5u_12_app_handoff_payload["selected_scoring_date"],
        errors="coerce",
    ).normalize()
    assert pd.notna(step5u_31_selected_scoring_date), "Selected scoring date is invalid."

    assert (
        step5u_31_available_scoring_date_min
        <= step5u_31_selected_scoring_date
        <= step5u_31_available_scoring_date_max
    ), "Selected scoring date is outside the available scoring range."

    assert (
        step5u_31_selected_scoring_date
        in step5u_31_available_dates_df["activity_date"].tolist()
    ), "Selected scoring date is not present in available scoring dates."

except Exception as exc:
    st.warning(
        "Brak poprawnej konfiguracji trybu scoringu albo wybranej daty. "
        "Sprawdź handoff aplikacyjny i zakres dostępnych dat."
    )
    st.error(str(exc))
    raise

# Build scoring mode summary
step5u_31_scoring_mode_summary_df = pd.DataFrame(
    [
        {
            "scoring_date_mode": step5u_31_scoring_date_mode,
            "selected_scoring_date": step5u_31_selected_scoring_date,
            "available_scoring_date_min": step5u_31_available_scoring_date_min,
            "available_scoring_date_max": step5u_31_available_scoring_date_max,
            "available_scoring_date_count": step5u_31_available_scoring_date_count,
            "null_semantics_row_count": int(step5u_31_null_semantics_df.shape[0]),
            "streamlit_runtime_available": step5u_31_streamlit_available,
            "scoring_mode_ready": 1,
        }
    ]
)

# Build supported modes table
step5u_31_supported_modes_df = pd.DataFrame(
    {
        "scoring_mode": step5u_22_supported_scoring_modes,
        "is_active_mode": [
            int(mode_name == step5u_31_scoring_date_mode)
            for mode_name in step5u_22_supported_scoring_modes
        ],
    }
)

# Build available date preview
step5u_31_available_date_preview_df = pd.concat(
    [
        step5u_31_available_dates_df.head(5),
        step5u_31_available_dates_df.tail(5),
    ],
    axis=0,
    ignore_index=True,
).drop_duplicates().reset_index(drop=True)

display(step5u_31_scoring_mode_summary_df)
display(step5u_31_supported_modes_df)
display(step5u_31_available_date_preview_df)

,scoring_date_mode,selected_scoring_date,available_scoring_date_min,available_scoring_date_max,available_scoring_date_count,null_semantics_row_count,streamlit_runtime_available,scoring_mode_ready
0,latest_available,2020-10-31,2017-05-02,2020-10-31,826,34,1,1


,scoring_mode,is_active_mode
0,latest_available,1
1,selected_date,0


,activity_date
0,2017-05-02
1,2017-05-03
2,2017-05-04
3,2017-05-05
4,2017-05-06
5,2020-10-27
6,2020-10-28
7,2020-10-29
8,2020-10-30
9,2020-10-31


In [7]:
# 5.3.2 Przygotowanie runtime wyboru daty scoringowej dla aplikacji

import pandas as pd

# Validate required objects from previous steps
required_objects = [
    "step5u_22_supported_scoring_modes",
    "step5u_22_selected_scoring_date",
    "step5u_31_scoring_date_mode",
    "step5u_31_available_dates_df",
    "step5u_31_available_scoring_date_min",
    "step5u_31_available_scoring_date_max",
]

missing_objects = [name for name in required_objects if name not in globals()]
assert not missing_objects, f"Missing required objects from previous steps: {missing_objects}"

# Define runtime helpers
step5u_32_available_dates_list = [
    pd.Timestamp(date_value).normalize()
    for date_value in step5u_31_available_dates_df["activity_date"].tolist()
]

assert len(step5u_32_available_dates_list) > 0, "Available scoring date list is empty."

step5u_32_latest_available_scoring_date = max(step5u_32_available_dates_list)
step5u_32_default_selected_date = pd.Timestamp(step5u_22_selected_scoring_date).normalize()

assert step5u_32_default_selected_date in step5u_32_available_dates_list, (
    "Default selected scoring date is not present in available scoring dates."
)

# Resolve active scoring mode for runtime
step5u_32_runtime_scoring_mode = str(step5u_31_scoring_date_mode)
assert step5u_32_runtime_scoring_mode in step5u_22_supported_scoring_modes, (
    f"Unsupported runtime scoring mode: {step5u_32_runtime_scoring_mode}"
)

# Resolve scoring date according to mode
try:
    if step5u_32_runtime_scoring_mode == "latest_available":
        step5u_32_resolved_scoring_date = step5u_32_latest_available_scoring_date
    elif step5u_32_runtime_scoring_mode == "selected_date":
        step5u_32_resolved_scoring_date = step5u_32_default_selected_date
    else:
        raise ValueError(f"Unsupported runtime scoring mode: {step5u_32_runtime_scoring_mode}")

    assert step5u_32_resolved_scoring_date in step5u_32_available_dates_list, (
        "Resolved scoring date is not present in available scoring dates."
    )

except Exception as exc:
    st.warning(
        "Nie udało się ustalić poprawnej daty scoringowej dla bieżącego trybu pracy aplikacji."
    )
    st.error(str(exc))
    raise

# Build widget-ready options
step5u_32_available_date_options_df = pd.DataFrame(
    {
        "activity_date": step5u_32_available_dates_list,
    }
)

step5u_32_available_date_options_df["date_label"] = (
    step5u_32_available_date_options_df["activity_date"]
    .dt.strftime("%Y-%m-%d")
)

step5u_32_default_selected_date_index = int(
    step5u_32_available_date_options_df.index[
        step5u_32_available_date_options_df["activity_date"] == step5u_32_default_selected_date
    ][0]
)

step5u_32_latest_available_date_index = int(
    step5u_32_available_date_options_df.index[
        step5u_32_available_date_options_df["activity_date"] == step5u_32_latest_available_scoring_date
    ][0]
)

step5u_32_runtime_selected_date_index = int(
    step5u_32_available_date_options_df.index[
        step5u_32_available_date_options_df["activity_date"] == step5u_32_resolved_scoring_date
    ][0]
)

# Build runtime mode summary
step5u_32_runtime_mode_summary_df = pd.DataFrame(
    [
        {
            "runtime_scoring_mode": step5u_32_runtime_scoring_mode,
            "default_selected_scoring_date": step5u_32_default_selected_date,
            "latest_available_scoring_date": step5u_32_latest_available_scoring_date,
            "resolved_scoring_date": step5u_32_resolved_scoring_date,
            "default_selected_date_index": step5u_32_default_selected_date_index,
            "latest_available_date_index": step5u_32_latest_available_date_index,
            "resolved_scoring_date_index": step5u_32_runtime_selected_date_index,
            "available_scoring_date_count": int(len(step5u_32_available_dates_list)),
            "runtime_date_selection_ready": 1,
        }
    ]
)

# Build UI mode option table
step5u_32_ui_mode_options_df = pd.DataFrame(
    {
        "scoring_mode": step5u_22_supported_scoring_modes,
        "ui_label": [
            "Latest available" if mode_name == "latest_available" else "Selected date"
            for mode_name in step5u_22_supported_scoring_modes
        ],
        "is_active_mode": [
            int(mode_name == step5u_32_runtime_scoring_mode)
            for mode_name in step5u_22_supported_scoring_modes
        ],
    }
)

display(step5u_32_runtime_mode_summary_df)
display(step5u_32_ui_mode_options_df)
display(step5u_32_available_date_options_df.head(10))

,runtime_scoring_mode,default_selected_scoring_date,latest_available_scoring_date,resolved_scoring_date,default_selected_date_index,latest_available_date_index,resolved_scoring_date_index,available_scoring_date_count,runtime_date_selection_ready
0,latest_available,2020-10-31,2020-10-31,2020-10-31,825,825,825,826,1


,scoring_mode,ui_label,is_active_mode
0,latest_available,Latest available,1
1,selected_date,Selected date,0


,activity_date,date_label
0,2017-05-02,2017-05-02
1,2017-05-03,2017-05-03
2,2017-05-04,2017-05-04
3,2017-05-05,2017-05-05
4,2017-05-06,2017-05-06
5,2017-05-07,2017-05-07
6,2017-05-08,2017-05-08
7,2017-05-09,2017-05-09
8,2017-05-10,2017-05-10
9,2017-05-11,2017-05-11


## Podsumowanie działu 5.3 Wybór trybu scoringu i daty scoringowej

**Kontekst inżynieryjny:** W dziale `5.3` zbudowano runtime logikę wyboru trybu scoringu i daty scoringowej dla aplikacji Streamlit. Najpierw odnaleziono i wczytano dwa kluczowe artefakty BLOKU 3 potrzebne do pracy aplikacji: `b3_13_model_ready_dataset.parquet` jako źródło rzeczywiście dostępnych dat scoringowych oraz `b3_13_null_semantics.parquet` jako kontrakt braków danych wymagany przy dalszej budowie wsadu inferencyjnego. Następnie odczytano aktywny tryb scoringu i zapisaną datę scoringu z handoffu aplikacyjnego, po czym zweryfikowano ich zgodność z listą dostępnych dat. W kolejnym kroku przygotowano runtime logikę rozstrzygania daty scoringowej dla dwóch trybów pracy aplikacji: `latest_available` oraz `selected_date`. Dodatkowo zbudowano gotowe do użycia w UI listy trybów i dat, tak aby późniejszy komponent Streamlit mógł bezpośrednio korzystać z poprawnie zweryfikowanych opcji wyboru.

**Interpretacja wyniku:** Z danych `model_ready` odczytano `826` unikalnych dat scoringowych, obejmujących zakres od `2017-05-02` do `2020-10-31`. Aktualny handoff aplikacyjny wskazuje tryb `latest_available`, a wybrana data scoringu została poprawnie zinterpretowana jako `2020-10-31`. Ta sama data została jednocześnie potwierdzona jako ostatnia dostępna data w zbiorze, co oznacza, że aktywny tryb oraz resolved scoring date są ze sobą w pełni zgodne. Wewnętrznie data `2020-10-31` znajduje się na pozycji `825` w liście `826` dostępnych dat, co dodatkowo potwierdza, że jest to ostatni dostępny batch scoringowy. Wspierane tryby scoringu zostały poprawnie utrzymane jako `latest_available` i `selected_date`, a dla interfejsu użytkownika przygotowano odpowiadające im etykiety `Latest available` i `Selected date`. Dodatkowo utworzono widget-ready listę dostępnych dat z czytelnymi etykietami tekstowymi w formacie `YYYY-MM-DD`, co umożliwi późniejsze bezpośrednie podpięcie tej listy do kontrolek wyboru w aplikacji.

**Znaczenie biznesowe:** Dział `5.3` rozwiązuje jeden z kluczowych problemów aplikacji użytkowej: sposób bezpiecznego i czytelnego wyboru daty scoringu. Dzięki temu użytkownik nie może wybrać daty spoza dostępnego zakresu, a aplikacja nie operuje na datach „wymyślonych”, tylko na rzeczywiście istniejących batchach scoringowych. To daje dwa istotne efekty biznesowe. Po pierwsze, tryb `latest_available` pozwala użytkownikowi szybko uruchomić scoring na najnowszym dostępnym dniu bez konieczności ręcznego ustawiania daty. Po drugie, tryb `selected_date` umożliwia historyczny replay scoringu dla dowolnej daty z dopuszczonego zakresu `2017-05-02 → 2020-10-31`, co zwiększa wartość demonstracyjną, testową i analityczną aplikacji. Jednocześnie obecność `null semantics` w runtime potwierdza, że dalsze kroki aplikacji będą zachowywać spójność z kontraktem inferencyjnym ustalonym wcześniej.

**Wniosek:** Dział `5.3` został domknięty poprawnie. BLOK 5 ma już gotową i zweryfikowaną logikę wyboru trybu scoringu, aktywnej daty scoringowej oraz pełnej listy dostępnych dat do wykorzystania w interfejsie Streamlit. Oznacza to, że można przejść do kolejnego etapu, czyli `5.4 Budowa wsadu i scoring w aplikacji`, mając pewność, że aplikacja startuje z poprawnie rozstrzygniętym trybem pracy i poprawnie określoną datą scoringu.

## 5.4 Budowa wsadu i scoring w aplikacji

In [8]:
# 5.4.1 Budowa runtime wsadu scoringowego dla wybranej daty

from pathlib import Path
import pandas as pd
import pyarrow.parquet as pq

# Define Streamlit fallback for notebook execution
try:
    import streamlit as st
except ImportError:
    class Step5u41StreamlitFallback:
        session_state = {}

        @staticmethod
        def warning(message):
            print(f"WARNING: {message}")

        @staticmethod
        def error(message):
            print(f"ERROR: {message}")

    st = Step5u41StreamlitFallback()

# Validate required objects from previous steps
required_objects = [
    "step5u_31_model_ready_path",
    "step5u_31_null_semantics_path",
    "step5u_22_feature_list",
    "step5u_22_feature_count",
    "step5u_32_resolved_scoring_date",
]

missing_objects = [name for name in required_objects if name not in globals()]
assert not missing_objects, f"Missing required objects from previous steps: {missing_objects}"

step5u_41_model_ready_path = Path(step5u_31_model_ready_path)
step5u_41_null_semantics_path = Path(step5u_31_null_semantics_path)
step5u_41_selected_scoring_date = pd.Timestamp(step5u_32_resolved_scoring_date).normalize()

assert step5u_41_model_ready_path.exists(), (
    f"Model-ready dataset does not exist: {step5u_41_model_ready_path}"
)
assert step5u_41_null_semantics_path.exists(), (
    f"Null-semantics dataset does not exist: {step5u_41_null_semantics_path}"
)

# Define runtime columns
step5u_41_key_columns = ["activity_date", "station_id"]
step5u_41_feature_columns = [str(feature_name) for feature_name in step5u_22_feature_list]

step5u_41_context_candidates = [
    "hub_flag",
    "persona",
    "microzone",
    "is_cold_start",
    "is_holiday",
    "is_business_free_day",
    "alert_hours_roll_sum_14",
    "alert_hours_lag_1",
    "consecutive_alert_days_before_t",
    "split_name",
]

# Read schema first and request only columns that really exist
step5u_41_model_ready_schema = pq.read_schema(step5u_41_model_ready_path)
step5u_41_model_ready_schema_columns = step5u_41_model_ready_schema.names

step5u_41_requested_columns = []
for column_name in step5u_41_key_columns + step5u_41_feature_columns + step5u_41_context_candidates:
    if column_name in step5u_41_model_ready_schema_columns and column_name not in step5u_41_requested_columns:
        step5u_41_requested_columns.append(column_name)

# Validate contract before loading data
step5u_41_missing_key_columns_in_schema = [
    column_name
    for column_name in step5u_41_key_columns
    if column_name not in step5u_41_model_ready_schema_columns
]
assert not step5u_41_missing_key_columns_in_schema, (
    f"Missing required key columns in model-ready schema: {step5u_41_missing_key_columns_in_schema}"
)

step5u_41_missing_feature_columns_in_schema = [
    column_name
    for column_name in step5u_41_feature_columns
    if column_name not in step5u_41_model_ready_schema_columns
]
assert not step5u_41_missing_feature_columns_in_schema, (
    f"Missing required feature columns in model-ready schema: {step5u_41_missing_feature_columns_in_schema}"
)

# Load only required existing columns from model-ready
try:
    step5u_41_model_ready_table = pq.read_table(
        step5u_41_model_ready_path,
        columns=step5u_41_requested_columns,
    )
    step5u_41_model_ready_df = step5u_41_model_ready_table.to_pandas()
except Exception as exc:
    st.warning("Could not load runtime scoring input from the model-ready dataset.")
    st.error(str(exc))
    raise

# Normalize keys
step5u_41_model_ready_df["activity_date"] = pd.to_datetime(
    step5u_41_model_ready_df["activity_date"],
    errors="coerce",
).dt.normalize()

step5u_41_model_ready_df["station_id"] = (
    step5u_41_model_ready_df["station_id"]
    .astype("string")
)

assert step5u_41_model_ready_df["activity_date"].notna().all(), (
    "Model-ready dataset contains invalid activity_date values."
)
assert step5u_41_model_ready_df["station_id"].notna().all(), (
    "Model-ready dataset contains missing station_id values."
)

# Slice batch for the selected scoring date
try:
    step5u_41_scoring_batch_source_df = (
        step5u_41_model_ready_df.loc[
            step5u_41_model_ready_df["activity_date"] == step5u_41_selected_scoring_date
        ]
        .copy()
        .reset_index(drop=True)
    )

    assert not step5u_41_scoring_batch_source_df.empty, (
        f"No scoring data found for selected date: {step5u_41_selected_scoring_date.date()}"
    )
except Exception as exc:
    st.warning("No historical data found for the selected date. Choose a different day.")
    st.error(str(exc))
    raise

# Validate feature contract
step5u_41_available_batch_columns = list(step5u_41_scoring_batch_source_df.columns)

step5u_41_missing_feature_columns = [
    column_name
    for column_name in step5u_41_feature_columns
    if column_name not in step5u_41_available_batch_columns
]
assert not step5u_41_missing_feature_columns, (
    f"Missing required feature columns in scoring batch: {step5u_41_missing_feature_columns}"
)

step5u_41_duplicate_feature_columns = int(
    pd.Series(step5u_41_feature_columns).duplicated().sum()
)
assert step5u_41_duplicate_feature_columns == 0, (
    "Feature list contains duplicate feature names."
)

# Build model matrix
step5u_41_model_input_df = (
    step5u_41_scoring_batch_source_df[step5u_41_feature_columns]
    .copy()
    .reset_index(drop=True)
)

assert step5u_41_model_input_df.shape[1] == step5u_22_feature_count, (
    "Runtime scoring matrix feature count is not consistent with the app contract."
)

assert list(step5u_41_model_input_df.columns) == step5u_41_feature_columns, (
    "Runtime scoring matrix feature order is not consistent with the app contract."
)

# Resolve available app context columns
step5u_41_app_context_columns = [
    column_name
    for column_name in step5u_41_context_candidates
    if column_name in step5u_41_scoring_batch_source_df.columns and column_name not in step5u_41_key_columns
]

# Validate null-semantics contract
step5u_41_null_semantics_df = pd.read_parquet(step5u_41_null_semantics_path)
assert not step5u_41_null_semantics_df.empty, "Null-semantics dataset is empty."

step5u_41_null_semantics_feature_key = next(
    (
        column_name
        for column_name in ["column_name", "feature_name", "feature"]
        if column_name in step5u_41_null_semantics_df.columns
    ),
    None,
)
assert step5u_41_null_semantics_feature_key is not None, (
    "Null-semantics dataset does not contain a recognized feature key column."
)

step5u_41_null_semantics_features = (
    step5u_41_null_semantics_df[step5u_41_null_semantics_feature_key]
    .astype("string")
    .tolist()
)

step5u_41_missing_null_semantics_features = [
    feature_name
    for feature_name in step5u_41_feature_columns
    if feature_name not in step5u_41_null_semantics_features
]
assert not step5u_41_missing_null_semantics_features, (
    f"Missing features in null-semantics contract: {step5u_41_missing_null_semantics_features}"
)

# Validate batch keys
step5u_41_duplicate_key_count = int(
    step5u_41_scoring_batch_source_df.duplicated(subset=step5u_41_key_columns).sum()
)
assert step5u_41_duplicate_key_count == 0, (
    "Runtime scoring batch contains duplicate day-station keys."
)

# Build scoring batch summary
step5u_41_scoring_batch_summary_df = pd.DataFrame(
    [
        {
            "selected_scoring_date": step5u_41_selected_scoring_date,
            "record_count": int(step5u_41_scoring_batch_source_df.shape[0]),
            "station_count": int(step5u_41_scoring_batch_source_df["station_id"].nunique()),
            "feature_count": int(step5u_41_model_input_df.shape[1]),
            "context_column_count": int(len(step5u_41_app_context_columns)),
            "missing_feature_count": int(len(step5u_41_missing_feature_columns)),
            "duplicate_key_count": step5u_41_duplicate_key_count,
            "scoring_batch_ready": 1,
        }
    ]
)

# Build null-semantics summary
step5u_41_null_semantics_summary_df = pd.DataFrame(
    [
        {
            "null_semantics_feature_key": step5u_41_null_semantics_feature_key,
            "expected_feature_count": int(len(step5u_41_feature_columns)),
            "available_null_semantics_feature_count": int(len(step5u_41_null_semantics_features)),
            "missing_null_semantics_feature_count": int(len(step5u_41_missing_null_semantics_features)),
            "null_semantics_contract_ready": 1,
        }
    ]
)

display(step5u_41_scoring_batch_summary_df)
display(step5u_41_null_semantics_summary_df)
display(step5u_41_scoring_batch_source_df[step5u_41_key_columns + step5u_41_app_context_columns].head(10))

,selected_scoring_date,record_count,station_count,feature_count,context_column_count,missing_feature_count,duplicate_key_count,scoring_batch_ready
0,2020-10-31,326,326,34,8,0,0,1


,null_semantics_feature_key,expected_feature_count,available_null_semantics_feature_count,missing_null_semantics_feature_count,null_semantics_contract_ready
0,column_name,34,34,0,1


,activity_date,station_id,hub_flag,is_cold_start,is_holiday,is_business_free_day,alert_hours_roll_sum_14,alert_hours_lag_1,consecutive_alert_days_before_t,split_name
0,2020-10-31,1,0,0.0,1,1,181.0,17.0,6.0,test
1,2020-10-31,10,0,0.0,1,1,158.0,20.0,3.0,test
2,2020-10-31,100,0,0.0,1,1,177.0,0.0,0.0,test
3,2020-10-31,101,1,0.0,1,1,248.0,17.0,12.0,test
4,2020-10-31,103,0,0.0,1,1,229.0,17.0,3.0,test
5,2020-10-31,104,0,0.0,1,1,178.0,13.0,1.0,test
6,2020-10-31,105,0,0.0,1,1,173.0,14.0,1.0,test
7,2020-10-31,106,0,0.0,1,1,217.0,18.0,11.0,test
8,2020-10-31,107,0,0.0,1,1,284.0,18.0,21.0,test
9,2020-10-31,108,0,0.0,1,1,174.0,17.0,6.0,test


In [9]:
# 5.4.2 Policzenie predicted_probability i predicted_label w runtime aplikacji

from datetime import datetime, timezone

import numpy as np
import pandas as pd

# Define Streamlit fallback for notebook execution
try:
    import streamlit as st
except ImportError:
    class Step5u42StreamlitFallback:
        session_state = {}

        @staticmethod
        def warning(message):
            print(f"WARNING: {message}")

        @staticmethod
        def error(message):
            print(f"ERROR: {message}")

    st = Step5u42StreamlitFallback()

# Validate required objects from previous steps
required_objects = [
    "step5u_21_model_object",
    "step5u_22_model_name",
    "step5u_22_model_release_tag",
    "step5u_22_decision_threshold",
    "step5u_41_selected_scoring_date",
    "step5u_41_scoring_batch_source_df",
    "step5u_41_model_input_df",
    "step5u_41_key_columns",
    "step5u_41_app_context_columns",
]

missing_objects = [name for name in required_objects if name not in globals()]
assert not missing_objects, f"Missing required objects from previous steps: {missing_objects}"

# Validate model runtime
assert hasattr(step5u_21_model_object, "predict_proba"), (
    "Loaded model does not expose predict_proba."
)
assert hasattr(step5u_21_model_object, "classes_"), (
    "Loaded model does not expose classes_."
)

step5u_42_model_classes = list(step5u_21_model_object.classes_)
assert len(step5u_42_model_classes) == 2, (
    "Runtime application scoring expects a binary classifier."
)

if 1 in step5u_42_model_classes:
    step5u_42_positive_class_label = 1
    step5u_42_positive_class_index = step5u_42_model_classes.index(1)
elif "1" in step5u_42_model_classes:
    step5u_42_positive_class_label = "1"
    step5u_42_positive_class_index = step5u_42_model_classes.index("1")
else:
    raise AssertionError(
        f"Positive class label 1 was not found in model classes: {step5u_42_model_classes}"
    )

# Run scoring
try:
    step5u_42_predict_proba_array = step5u_21_model_object.predict_proba(
        step5u_41_model_input_df
    )
except Exception as exc:
    st.warning("Runtime scoring could not be completed for the selected scoring batch.")
    st.error(str(exc))
    raise

assert step5u_42_predict_proba_array.ndim == 2, (
    "predict_proba output must be a 2D array."
)
assert step5u_42_predict_proba_array.shape[0] == step5u_41_model_input_df.shape[0], (
    "Probability row count is not consistent with the runtime scoring input."
)
assert step5u_42_predict_proba_array.shape[1] == len(step5u_42_model_classes), (
    "Probability column count is not consistent with model classes."
)

step5u_42_predicted_probability = step5u_42_predict_proba_array[:, step5u_42_positive_class_index]
step5u_42_predicted_probability = np.asarray(
    step5u_42_predicted_probability,
    dtype="float64",
)

assert np.isfinite(step5u_42_predicted_probability).all(), (
    "Predicted probabilities contain non-finite values."
)
assert (
    (step5u_42_predicted_probability >= 0.0)
    & (step5u_42_predicted_probability <= 1.0)
).all(), (
    "Predicted probabilities must be in the range [0, 1]."
)

step5u_42_applied_threshold = float(step5u_22_decision_threshold)
step5u_42_predicted_label = (
    step5u_42_predicted_probability >= step5u_42_applied_threshold
).astype("int8")

step5u_42_scoring_timestamp = datetime.now(timezone.utc).replace(microsecond=0).isoformat()

# Build runtime scoring output
step5u_42_runtime_predictions_df = (
    step5u_41_scoring_batch_source_df[
        step5u_41_key_columns + step5u_41_app_context_columns
    ]
    .copy()
    .reset_index(drop=True)
)

step5u_42_runtime_predictions_df["predicted_probability"] = (
    step5u_42_predicted_probability.astype("float32")
)
step5u_42_runtime_predictions_df["predicted_label"] = (
    step5u_42_predicted_label.astype("int8")
)
step5u_42_runtime_predictions_df["scoring_timestamp"] = step5u_42_scoring_timestamp
step5u_42_runtime_predictions_df["model_release_tag"] = step5u_22_model_release_tag
step5u_42_runtime_predictions_df["applied_threshold"] = step5u_42_applied_threshold

step5u_42_output_columns = [
    "activity_date",
    "station_id",
    "predicted_probability",
    "predicted_label",
    "scoring_timestamp",
    "model_release_tag",
    "applied_threshold",
] + step5u_41_app_context_columns

step5u_42_runtime_predictions_df = (
    step5u_42_runtime_predictions_df[step5u_42_output_columns]
    .sort_values(["predicted_probability", "station_id"], ascending=[False, True])
    .reset_index(drop=True)
)

# Validate runtime output
step5u_42_duplicate_prediction_key_count = int(
    step5u_42_runtime_predictions_df.duplicated(
        subset=["activity_date", "station_id", "scoring_timestamp"]
    ).sum()
)
assert step5u_42_duplicate_prediction_key_count == 0, (
    "Runtime predictions contain duplicate prediction keys."
)

assert step5u_42_runtime_predictions_df["activity_date"].nunique() == 1, (
    "Runtime predictions must contain exactly one scoring date."
)
assert pd.Timestamp(step5u_42_runtime_predictions_df["activity_date"].iloc[0]).normalize() == (
    pd.Timestamp(step5u_41_selected_scoring_date).normalize()
), (
    "Runtime prediction date is not consistent with the selected scoring date."
)

# Persist runtime scoring output in session state
st.session_state["scored_data"] = step5u_42_runtime_predictions_df.copy()

# Build runtime scoring summary
step5u_42_runtime_scoring_summary_df = pd.DataFrame(
    [
        {
            "model_name": step5u_22_model_name,
            "model_release_tag": step5u_22_model_release_tag,
            "selected_scoring_date": step5u_41_selected_scoring_date,
            "record_count": int(step5u_42_runtime_predictions_df.shape[0]),
            "positive_prediction_count": int(step5u_42_runtime_predictions_df["predicted_label"].sum()),
            "positive_prediction_share": float(step5u_42_runtime_predictions_df["predicted_label"].mean()),
            "probability_min": float(step5u_42_runtime_predictions_df["predicted_probability"].min()),
            "probability_mean": float(step5u_42_runtime_predictions_df["predicted_probability"].mean()),
            "probability_max": float(step5u_42_runtime_predictions_df["predicted_probability"].max()),
            "applied_threshold": step5u_42_applied_threshold,
            "session_state_scored_data_ready": int("scored_data" in st.session_state),
            "runtime_scoring_ready": 1,
        }
    ]
)

display(step5u_42_runtime_scoring_summary_df)
display(step5u_42_runtime_predictions_df.head(10))

2026-04-22 13:12:29.042 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-04-22 13:12:29.044 WARNING streamlit.runtime.state.session_state_proxy: Session state does not function when running a script without `streamlit run`
2026-04-22 13:12:29.046 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-04-22 13:12:29.052 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.


,model_name,model_release_tag,selected_scoring_date,record_count,positive_prediction_count,positive_prediction_share,probability_min,probability_mean,probability_max,applied_threshold,session_state_scored_data_ready,runtime_scoring_ready
0,lightgbm_tuned,b4_15,2020-10-31,326,303,0.929448,0.082192,0.73389,0.951512,0.55,1,1


,activity_date,station_id,predicted_probability,predicted_label,scoring_timestamp,model_release_tag,applied_threshold,hub_flag,is_cold_start,is_holiday,is_business_free_day,alert_hours_roll_sum_14,alert_hours_lag_1,consecutive_alert_days_before_t,split_name
0,2020-10-31,201,0.951512,1,2026-04-22T11:12:29+00:00,b4_15,0.55,0,0.0,1,1,272.0,18.0,69.0,test
1,2020-10-31,16,0.932096,1,2026-04-22T11:12:29+00:00,b4_15,0.55,0,0.0,1,1,275.0,17.0,48.0,test
2,2020-10-31,5,0.927936,1,2026-04-22T11:12:29+00:00,b4_15,0.55,0,0.0,1,1,241.0,17.0,2.0,test
3,2020-10-31,64,0.927556,1,2026-04-22T11:12:29+00:00,b4_15,0.55,1,0.0,1,1,260.0,18.0,11.0,test
4,2020-10-31,113,0.925940,1,2026-04-22T11:12:29+00:00,b4_15,0.55,1,0.0,1,1,283.0,7.0,61.0,test
5,2020-10-31,101,0.923846,1,2026-04-22T11:12:29+00:00,b4_15,0.55,1,0.0,1,1,248.0,17.0,12.0,test
6,2020-10-31,547,0.923610,1,2026-04-22T11:12:29+00:00,b4_15,0.55,1,0.0,1,1,291.0,23.0,36.0,test
7,2020-10-31,58,0.920777,1,2026-04-22T11:12:29+00:00,b4_15,0.55,0,0.0,1,1,245.0,19.0,55.0,test
8,2020-10-31,63,0.908855,1,2026-04-22T11:12:29+00:00,b4_15,0.55,0,0.0,1,1,257.0,23.0,40.0,test
9,2020-10-31,4,0.903446,1,2026-04-22T11:12:29+00:00,b4_15,0.55,0,0.0,1,1,252.0,17.0,11.0,test


In [10]:
# 5.4.3 Domknięcie runtime scoringu i walidacja scored_data dla aplikacji

import pandas as pd

# Define Streamlit fallback for notebook execution
try:
    import streamlit as st
    step5u_43_streamlit_runtime_available = 1
except ImportError:
    step5u_43_streamlit_runtime_available = 0

    class Step5u43StreamlitFallback:
        session_state = {}

        @staticmethod
        def warning(message):
            print(f"WARNING: {message}")

        @staticmethod
        def error(message):
            print(f"ERROR: {message}")

    st = Step5u43StreamlitFallback()

# Validate required objects from previous steps
required_objects = [
    "step5u_22_model_name",
    "step5u_22_model_release_tag",
    "step5u_22_decision_threshold",
    "step5u_41_selected_scoring_date",
    "step5u_41_app_context_columns",
    "step5u_42_runtime_predictions_df",
    "step5u_42_runtime_scoring_summary_df",
]

missing_objects = [name for name in required_objects if name not in globals()]
assert not missing_objects, f"Missing required objects from previous steps: {missing_objects}"

# Resolve scored_data source
if "scored_data" in st.session_state:
    step5u_43_scored_data_df = (
        st.session_state["scored_data"]
        .copy()
        .reset_index(drop=True)
    )
    step5u_43_scored_data_source = "streamlit_session_state"
else:
    step5u_43_scored_data_df = (
        step5u_42_runtime_predictions_df
        .copy()
        .reset_index(drop=True)
    )
    st.session_state["scored_data"] = step5u_43_scored_data_df.copy()
    step5u_43_scored_data_source = "runtime_predictions_fallback"

# Define expected runtime output columns
step5u_43_base_output_columns = [
    "activity_date",
    "station_id",
    "predicted_probability",
    "predicted_label",
    "scoring_timestamp",
    "model_release_tag",
    "applied_threshold",
]
step5u_43_expected_output_columns = (
    step5u_43_base_output_columns + step5u_41_app_context_columns
)

assert list(step5u_43_scored_data_df.columns) == step5u_43_expected_output_columns, (
    "Runtime scored_data columns are not consistent with the expected app output contract."
)

# Normalize key fields
step5u_43_scored_data_df["activity_date"] = pd.to_datetime(
    step5u_43_scored_data_df["activity_date"],
    errors="coerce",
).dt.normalize()

step5u_43_scored_data_df["station_id"] = (
    step5u_43_scored_data_df["station_id"]
    .astype("string")
)

assert step5u_43_scored_data_df["activity_date"].notna().all(), (
    "Runtime scored_data contains invalid activity_date values."
)
assert step5u_43_scored_data_df["station_id"].notna().all(), (
    "Runtime scored_data contains missing station_id values."
)

assert step5u_43_scored_data_df["predicted_probability"].between(0.0, 1.0).all(), (
    "Runtime scored_data contains probabilities outside [0, 1]."
)
assert step5u_43_scored_data_df["predicted_label"].isin([0, 1]).all(), (
    "Runtime scored_data contains invalid predicted_label values."
)
assert (step5u_43_scored_data_df["model_release_tag"] == step5u_22_model_release_tag).all(), (
    "Runtime scored_data contains inconsistent model_release_tag values."
)
assert (step5u_43_scored_data_df["applied_threshold"] == float(step5u_22_decision_threshold)).all(), (
    "Runtime scored_data contains inconsistent applied_threshold values."
)

step5u_43_duplicate_key_count = int(
    step5u_43_scored_data_df.duplicated(
        subset=["activity_date", "station_id", "scoring_timestamp"]
    ).sum()
)
assert step5u_43_duplicate_key_count == 0, (
    "Runtime scored_data contains duplicate prediction keys."
)

assert step5u_43_scored_data_df["activity_date"].nunique() == 1, (
    "Runtime scored_data must contain exactly one scoring date."
)
assert pd.Timestamp(step5u_43_scored_data_df["activity_date"].iloc[0]).normalize() == (
    pd.Timestamp(step5u_41_selected_scoring_date).normalize()
), (
    "Runtime scored_data date is not consistent with the selected scoring date."
)

# Validate sort contract
step5u_43_sorted_reference_df = (
    step5u_43_scored_data_df.sort_values(
        ["predicted_probability", "station_id"],
        ascending=[False, True],
    )
    .reset_index(drop=True)
)

assert step5u_43_scored_data_df.equals(step5u_43_sorted_reference_df), (
    "Runtime scored_data is not sorted according to the app output contract."
)

# Build runtime output contract summary
step5u_43_runtime_output_contract_df = pd.DataFrame(
    [
        {
            "model_name": step5u_22_model_name,
            "model_release_tag": step5u_22_model_release_tag,
            "selected_scoring_date": step5u_41_selected_scoring_date,
            "record_count": int(step5u_43_scored_data_df.shape[0]),
            "output_column_count": int(step5u_43_scored_data_df.shape[1]),
            "context_column_count": int(len(step5u_41_app_context_columns)),
            "duplicate_key_count": step5u_43_duplicate_key_count,
            "streamlit_runtime_available": step5u_43_streamlit_runtime_available,
            "scored_data_source": step5u_43_scored_data_source,
            "session_state_scored_data_ready": int("scored_data" in st.session_state),
            "runtime_output_contract_ready": 1,
        }
    ]
)

# Build filter readiness summary for the next screen
step5u_43_filter_readiness_df = pd.DataFrame(
    [
        {
            "station_id_filter_ready": int("station_id" in step5u_43_scored_data_df.columns),
            "hub_flag_filter_ready": int("hub_flag" in step5u_43_scored_data_df.columns),
            "predicted_label_filter_ready": int("predicted_label" in step5u_43_scored_data_df.columns),
            "is_cold_start_filter_ready": int("is_cold_start" in step5u_43_scored_data_df.columns),
            "is_holiday_filter_ready": int("is_holiday" in step5u_43_scored_data_df.columns),
            "filter_layer_ready": 1,
        }
    ]
)

# Build KPI preview summary for the next screen
step5u_43_kpi_preview_df = pd.DataFrame(
    [
        {
            "station_count": int(step5u_43_scored_data_df["station_id"].nunique()),
            "high_risk_station_count": int(step5u_43_scored_data_df["predicted_label"].sum()),
            "high_risk_station_share": float(step5u_43_scored_data_df["predicted_label"].mean()),
            "cold_start_in_high_risk_count": int(
                step5u_43_scored_data_df.loc[
                    (step5u_43_scored_data_df["predicted_label"] == 1)
                    & (step5u_43_scored_data_df["is_cold_start"] == 1)
                ].shape[0]
            ) if "is_cold_start" in step5u_43_scored_data_df.columns else pd.NA,
            "holiday_in_high_risk_count": int(
                step5u_43_scored_data_df.loc[
                    (step5u_43_scored_data_df["predicted_label"] == 1)
                    & (step5u_43_scored_data_df["is_holiday"] == 1)
                ].shape[0]
            ) if "is_holiday" in step5u_43_scored_data_df.columns else pd.NA,
            "kpi_layer_ready": 1,
        }
    ]
)

display(step5u_43_runtime_output_contract_df)
display(step5u_43_filter_readiness_df)
display(step5u_43_kpi_preview_df)

2026-04-22 13:12:29.176 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-04-22 13:12:29.178 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-04-22 13:12:29.221 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.


,model_name,model_release_tag,selected_scoring_date,record_count,output_column_count,context_column_count,duplicate_key_count,streamlit_runtime_available,scored_data_source,session_state_scored_data_ready,runtime_output_contract_ready
0,lightgbm_tuned,b4_15,2020-10-31,326,15,8,0,1,streamlit_session_state,1,1


,station_id_filter_ready,hub_flag_filter_ready,predicted_label_filter_ready,is_cold_start_filter_ready,is_holiday_filter_ready,filter_layer_ready
0,1,1,1,1,1,1


,station_count,high_risk_station_count,high_risk_station_share,cold_start_in_high_risk_count,holiday_in_high_risk_count,kpi_layer_ready
0,326,303,0.929448,0,303,1


## Podsumowanie działu 5.4 Budowa wsadu i scoring w aplikacji

**Kontekst inżynieryjny:** W dziale `5.4` zbudowano runtime scoring aplikacji na podstawie tej samej logiki inferencyjnej, która została wcześniej zweryfikowana w BLOKU 4. Najpierw przygotowano runtime batch scoringowy dla wybranej daty, wykorzystując `b3_13_model_ready_dataset.parquet` jako źródło danych dziennych oraz `b3_13_null_semantics.parquet` jako kontrakt logiki braków. Przy odczycie danych zastosowano defensywną walidację schematu pliku, tak aby wczytywać tylko te kolumny kontekstowe, które rzeczywiście istnieją w model-ready dataset, przy jednoczesnym zachowaniu obowiązkowości kluczy i wszystkich cech modelowych. Następnie zbudowano macierz modelową zgodną z kontraktem inferencyjnym, zweryfikowano kompletność cech, brak duplikatów klucza `activity_date + station_id` oraz pełne pokrycie `null semantics`. W kolejnym kroku uruchomiono scoring runtime przy użyciu załadowanego modelu `lightgbm_tuned`, policzono `predicted_probability`, utworzono `predicted_label` przy sztywnym progu `0.55`, dodano metadane traceability, a wynik runtime został w środowisku notebookowym poprawnie zmaterializowany z step5u_42_runtime_predictions_df i osadzony w warstwie st.session_state["scored_data"], co przygotowuje aplikację do dalszego filtrowania i prezentacji bez ponownego uruchamiania modelu. Na końcu domknięto warstwę runtime output, walidując kontrakt wyjścia, gotowość filtrów oraz gotowość KPI dla ekranu wyników.

**Interpretacja wyniku:** Runtime batch scoringowy został poprawnie zbudowany dla daty `2020-10-31` i obejmuje `326` rekordów oraz `326` unikalnych stacji. Macierz modelowa zawiera komplet `34` cech wymaganych przez model, nie występują żadne brakujące cechy (`missing_feature_count = 0`) ani duplikaty klucza (`duplicate_key_count = 0`). Dla warstwy użytkowej zidentyfikowano `8` kolumn kontekstowych, obejmujących m.in. `hub_flag`, `is_cold_start`, `is_holiday`, `is_business_free_day`, `alert_hours_roll_sum_14`, `alert_hours_lag_1`, `consecutive_alert_days_before_t` oraz `split_name`. Kontrakt `null semantics` jest w pełni zgodny z modelem: oczekiwane `34` cechy są obecne w kontrakcie i żadna cecha nie jest brakująca (`missing_null_semantics_feature_count = 0`). Runtime scoring zwrócił `326` predykcji, z czego `303` otrzymały etykietę dodatnią, co odpowiada udziałowi około `92.94%` batcha. Zakres prawdopodobieństw wynosi od około `0.0822` do `0.9515`, przy średniej około `0.7339`. Warstwa `scored_data` została poprawnie przygotowana do dalszego użycia, ma `15` kolumn, nie zawiera duplikatów i jest poprawnie posortowana malejąco po `predicted_probability`. Dodatkowo potwierdzono gotowość wszystkich kluczowych filtrów (`station_id`, `hub_flag`, `predicted_label`, `is_cold_start`, `is_holiday`) oraz gotowość warstwy KPI. Podgląd KPI pokazuje `326` stacji w batchu, `303` stacje wysokiego ryzyka, udział wysokiego ryzyka około `92.94%`, `0` cold startów wśród stacji wysokiego ryzyka oraz `303` rekordy wysokiego ryzyka przypadające na dzień oznaczony jako świąteczny / wolny.

**Znaczenie biznesowe:** Dział `5.4` przenosi aplikację z poziomu samej architektury i metadanych na poziom rzeczywistego działania modelu w runtime. Oznacza to, że aplikacja nie tylko zna model, próg i zakres dat, ale potrafi już zbudować poprawny wsad dla wybranego dnia, zwalidować kontrakt inferencyjny, policzyć wynik i przygotować dane bezpośrednio do prezentacji użytkownikowi. Dzięki zapisowi do `st.session_state["scored_data"]` kolejne akcje w aplikacji, takie jak filtrowanie, sortowanie czy eksport, nie wymagają ponownego uruchamiania modelu, co jest kluczowe z punktu widzenia wydajności i komfortu użytkowania. Dodatkowo obecność kontekstu operacyjnego oraz lekkiego explainability sprawia, że wynik modelu może być prezentowany nie jako „czarna skrzynka”, lecz jako ranking ryzyka z podstawowym uzasadnieniem.

**Wniosek:** Dział `5.4` został domknięty poprawnie. BLOK 5 ma już działający runtime scoring oparty na gotowym modelu, poprawnie zbudowany batch inferencyjny dla wybranej daty, zwalidowany kontrakt wyjścia oraz gotową warstwę `scored_data` do użycia w aplikacji. Oznacza to, że można przejść do kolejnego etapu, czyli `5.5 Prezentacja wyników i filtrów użytkowych`, mając pewność, że frontend będzie pracował na poprawnie policzonym i zwalidowanym wyniku runtime.

## Blok 5 Aplikacja Streamlit

## aplikacja_dzien_stacja.py  http://localhost:8501/